### PR TITLE
Make sessions persist to fix invalid state

### PR DESCRIPTION
**Pull request by bkildow**
See https://github.com/auth0/auth0-drupal/issues/139.

### Checklist

* [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [x] I have read the [Auth0 code of conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
* [x] I ran the PHPCS Drupal coding standards

### CODE REVIEW

**Comment by joshcanhelp** [comment]
@bkildow - Thanks for this and great timing, we're just testing a release now. 

Can you provide steps for how to test this failing before the change and succeeding after? That would be helpful on my end to validate this.

**Comment by bkildow** [comment]
@joshcanhelp, awesome, good to hear about timing.

It's failing with an exception thrown here: https://github.com/auth0/auth0-drupal/blob/aead0f9fd3ef9c3cf5265274b8a7864b911c89f3/src/Controller/AuthController.php#L440. I'm using the universal login page, if it makes a difference. The exception message is "Invalid state". On the Drupal side, the user doesn't get logged in.

After the change, no exception should be thrown (logged), and Drupal user should be logged in.

**Comment by joshcanhelp** [comment]
@bkildow - Thanks for that but I'm looking for what steps you're taking to generate this exception. Basic login with ULP or embedded should be working as expected so I'm guessing there is an extra step or configuration that I'm not testing? I just want to make sure we're covering as many testing cases as possible/is reasonable. 

One thing that might be causing the issue here is that the PHP SDK that this module relies on is doing its own work with the session, possibly in ways that are non-standard for Drupal. Still, I'm able to log in successfully either way with the module in its current state.

**Comment by bkildow** [comment]
@joshcanhelp, ugh, I just tried on a fresh 8.7 install, and I can't reproduce the issue. I'm wondering if I'm running into a core issue that got fixed between the version I'm currently on, 8.7-alpha1, and yesterday's 8.7 stable release (maybe this: https://www.drupal.org/project/drupal/issues/3045349, not sure). I plan on updating to the 8.7 release as soon as the upstream distro I'm on cuts a release with this, which I'm guessing will fix my issue.

**Comment by joshcanhelp** [comment]
Since the `use` declaration didn't change, maybe what's being injected is the same and this was just a temporary core issue like you mentioned? Might be good to see what's being injected before and after the change and see if that does anything different, as most of the changes here are just changing the var name. 

Let me know if I can help test, happy to do so!

**Comment by joshcanhelp** [comment]
@bkildow - Any news on this one? I'm testing in preparation to release the module this week. Let me know if I can help at all.

**Comment by joshcanhelp** [comment]
I just tested this change out and both seem to work as expected. I'm not seeing a "bare" PHP session cookie so I assume that it's using Drupal session management properly. The currently-injected `Drupal\Core\Session\SessionManagerInterface` is using the Symfony component under the hood so I don't think this change affects anything. 

I'll leave this open for another couple of days. Let me know if you find anything on your end.

**Comment by joshcanhelp** [comment]
Tested with 8.7.0 and everything works well with the module in its current state.

**Comment by joshcanhelp** [comment]
Going to close this for now. Please let me know if this crops up again.

**Comment by cachesclay** [comment]
Was getting same errors, preventing log in, and this patch fixed it. My site is decoupled, which could be leading to some of the session issues, but I was about to pull apart the module until I saw this (also didn't realize the issue queue had been moved over here from d.o, or I would have searched here first).

**Comment by joshcanhelp** [comment]
@cachesclay - Thank you for the report here! Can you describe how to recreate the issue so I can test it out on my end? I'm not sure what you mean by "My site is decoupled."

**Comment by cachesclay** [comment]
@joshcanhelp our drupal site is being utilized as a backend, and completely decoupled from the front end presentation layer, where we are using react (For more info on decouple: https://dri.es/how-to-decouple-drupal-in-2019). We have a very complicated site setup, so I'm not sure the best way to replicate, tbh.

**Comment by joshcanhelp** [comment]
@cachesclay - It works either way for me so if this will work for your setup and there are no expected issues with others, I have no problem merging this in.

**Comment by joshcanhelp** [comment]
@rob3000 - If you have a chance, can you take a quick look at this? I'm not totally familiar with the change here but it's working on my test install and sounds like it fixes an issue. Thank you!

File: src/Controller/AuthController.php [hunk 1]
```diff
@@ -224,7 +225,7 @@ public function __construct(
   public static function create(ContainerInterface $container) {
     return new static(
         $container->get('user.private_tempstore'),
-        $container->get('session_manager'),
+        $container->get('session'),
```
Comment (rob3000) [inline src/Controller/AuthController.php]: @bkildow curious as to the change from session_manager, from what i can see the session_manager is a Drupal wrapper around the Symfony session service which is what this was changed to. Assuming this was needed for the `migrate` method? If so the SessionManager has `regenerate` cc @joshcanhelp

Comment (cachesclay) [reply src/Controller/AuthController.php]: +1 on using session_manager migrate

Comment (joshcanhelp) [reply src/Controller/AuthController.php]: @cachesclay - You can confirm that this works on your decoupled install? Just tried that locally and it works fine.

Comment (cachesclay) [reply src/Controller/AuthController.php]: It should be the same, it's the same method, just being wrapped. Let me push up to my dev server and see.

Comment (cachesclay) [reply src/Controller/AuthController.php]: yep, can confirm that this fixes my issue:
`diff --git a/src/Controller/AuthController.php b/src/Controller/AuthController.php
index c7827122..d0676cf7 100644
--- a/src/Controller/AuthController.php
+++ b/src/Controller/AuthController.php
@@ -319,7 +319,7 @@ protected function getNonce($returnTo) {
     // we don't actually start it!
     if (!$this->sessionManager->isStarted() && !isset($_SESSION['auth0_is_session_started'])) {
       $_SESSION['auth0_is_session_started'] = 'yes';
-      $this->sessionManager->start();
+      $this->sessionManager->regenerate();
     }
 
     $sessionStateHandler = new SessionStateHandler(new SessionStore());
`

File: src/Controller/AuthController.php [hunk 2]
```diff
@@ -317,9 +318,9 @@ public function logout() {
   protected function getNonce($returnTo) {
     // Have to start the session after putting something into the session, or
     // we don't actually start it!
-    if (!$this->sessionManager->isStarted() && !isset($_SESSION['auth0_is_session_started'])) {
+    if (!$this->session->isStarted() && !isset($_SESSION['auth0_is_session_started'])) {
       $_SESSION['auth0_is_session_started'] = 'yes';
-      $this->sessionManager->start();
+      $this->session->migrate();
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: @bkildow - It looks like we can revert all of these changes and switch `$this->sessionManager->start()` to `$this->sessionManager->regenerate()`

Comment (bkildow) [reply src/Controller/AuthController.php]: @joshcanhelp that makes sense, I have update the PR accordingly.

**Pull request merged**